File: algorithms/2_medium/M/03927_minimize_array_sum_using_divisible_replacements/solution.c
long long minArraySum(const int* nums, const int len)
{
    assert(len > 0 && nums != NULL);
    bool any_one = false;
    for (int i = 0; i < len; i += 1) {
        if (nums[i] == 1) {
            any_one = true;
            break;
        }
    }
    if (any_one) {
        return len;
    }

    int max = INT_MIN;
    for (int i = 0; i < len; i += 1) {
        assert(nums[i] >= 1 && nums[i] <= 100000);
        if (nums[i] > max) {
            max = nums[i];
        }
    }
    int* freqs = calloc((size_t)(max + 1), sizeof(*freqs));
    assert(freqs != NULL);
    for (int i = 0; i < len; i += 1) {
        freqs[nums[i]] += 1;
    }
    for (int i = 0; i <= max; i += 1) {
        if (freqs[i] != 0) {
            int curr = i + i;
            while (curr <= max) {
                freqs[i] += freqs[curr];
                freqs[curr] = 0;
                curr += i;
            }
        }
    }

    int64_t res = 0;
    for (int i = 1; i <= max; i += 1) {
        res += i * (int64_t)freqs[i];
    }

/* cleanup: */
    free(freqs);
    return res;
}

File: algorithms/2_medium/M/03927_minimize_array_sum_using_divisible_replacements/solution.rs
impl Solution {
    pub fn min_array_sum(nums: Vec<i32>) -> i64 {
        assert!(nums.len() > 0);
        debug_assert!(nums.iter().all(|x| *x >= 1 && *x <= 100_000));
        if nums.iter().any(|x| *x == 1) {
            return nums.len() as i64;
        }
        let max = *nums.iter().max().unwrap();
        let mut freqs = vec![0_i32; (max + 1) as usize];
        for num in nums {
            freqs[num as usize] += 1;
        }
        for i in 1..=max {
            if freqs[i as usize] != 0 {
                let mut curr = i + i;
                while curr <= max {
                    freqs[i as usize] += freqs[curr as usize];
                    freqs[curr as usize] = 0;
                    curr += i;
                }
            }
        }

        let mut res = 0_i64;
        for (i_usize, freq_i32) in freqs.into_iter().enumerate() {
            let freq = freq_i32 as i64;
            let i = i_usize as i64;
            res += i * freq;
        }

        res
    }
}
